base/broadcast.jl

```diff
@@ -1385,4 +1385,50 @@ function Base.show(io::IO, op::BroadcastFunction)
 end
 Base.show(io::IO, ::MIME"text/plain", op::BroadcastFunction) = show(io, op)
 
+struct IMArrayStyle <: Broadcast.AbstractArrayStyle{Any} end
+BroadcastStyle(::Type{<:Core.ImmutableArray}) = IMArrayStyle()
+
+#similar has to return mutable array 
+function Base.similar(bc::Broadcasted{IMArrayStyle}, ::Type{ElType}) where ElType
+    similar(Array{ElType}, axes(bc))
+end
+
+#but we define im_similar for giving a similar immutable array
+function im_similar(bc::Broadcasted{IMArrayStyle}, ::Type{ElType}) where ElType
+    Core.ImmutableArray(similar(bc, ElType))
+end
+
+#for now, make IMArrayStyle beat others to keep immutability
+BroadcastStyle(a::IMArrayStyle, ::DefaultArrayStyle) = a
+BroadcastStyle(a::IMArrayStyle, ::AbstractArrayStyle) = a
+
+@inline function copy(bc::Broadcasted{IMArrayStyle}) where {IMArrayStyle}
+    ElType = combine_eltypes(bc.f, bc.args)
+    if Base.isconcretetype(ElType)
+        # This seems like a mess of copying...
+        return Core.ImmutableArray(copyto!(similar(bc, ElType), bc))
+    end
+    # When ElType is not concrete, use narrowing. Use the first output
+    # value to determine the starting output eltype; copyto_nonleaf!
+    # will widen `dest` as needed to accommodate later values.
+    bc′ = preprocess(nothing, bc)
+    iter = eachindex(bc′)
+    y = iterate(iter)
+    if y === nothing
+        # if empty, take the ElType at face value
+        return im_similar(bc′, ElType)
+    end
+    # Initialize using the first value
+    I, state = y
+    @inbounds val = bc′[I]
+    dest = similar(bc′, typeof(val)) #need mutable here
+    @inbounds dest[I] = val
+    # Now handle the remaining values
+    # The typeassert gives inference a helping hand on the element type and dimensionality
+    # (work-around for #28382)
+    ElType′ = ElType === Union{} ? Any : ElType <: Type ? Type : ElType
+    RT = dest isa AbstractArray ? AbstractArray{<:ElType′, ndims(dest)} : Any
```

Comment:
Not directly related to this PR, but might be a good opportunity to check whether this still holds true with the recent inference improvements.